File: python/leetcode/221.py
class Solution:
    def maximalSquare(self, matrix):
        if not matrix: return 0
        m, n = len(matrix), len(matrix[0])
        dp = [[0] * n for x in range(m)]
        ans = 0
        for x in range(m):
            for y in range(n):
                dp[x][y] = int(matrix[x][y])
                if x and y and dp[x][y]:
                    dp[x][y] = min(dp[x - 1][y - 1], dp[x][y - 1], dp[x - 1][y]) + 1
                ans = max(ans, dp[x][y])
        return ans * ans


class Solution:
    def maximalSquare(self, matrix):
        m = len(matrix)
        n = len(matrix[0]) if m > 0 else 0
        dp = [0 for i in range(n+1)]
        ans, prev = 0, 0
        for i in range(1, m+1):
            for j in range(1, n+1):
                temp = dp[j]
                if matrix[i - 1][j - 1] == '1':
                    dp[j] = min(dp[j - 1], prev, dp[j]) + 1
                    ans = max(ans, dp[j])
                else:
                    dp[j] = 0
                prev = temp
        return ans * ans

File: cpp/leetcode/221.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if(matrix.empty() || matrix[0].empty()) return 0;
        int m = matrix.size(), n = matrix[0].size();
        vector<vector<int>> dp(m + 1, vector<int>(n+1, 0));
        int ans = 0;
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; ++j){
                if(matrix[i-1][j-1] == '1'){
                    dp[i][j] = 1 + min(dp[i-1][j-1], min(dp[i-1][j], dp[i][j-1]));
                }
                ans = max(dp[i][j], ans);
            }
        }
        return ans * ans;
    }
};


class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if(matrix.empty() || matrix[0].empty()) return 0;
        int m = matrix.size(), n = matrix[0].size();
        vector<int> dp(n+1, 0);
        int ans = 0, pre = 0;
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; ++j){
                if(matrix[i-1][j-1] == '1'){
                    int tmp = dp[j];
                    dp[j] = 1 + min(dp[j-1], min(dp[j], pre));
                    pre = tmp;
                } else {
                    dp[j] = 0;
                }
                ans = max(dp[j], ans);
            }
        }
        return ans * ans;
    }
};